File: src_all/libs_src/MobileIMSDKServerX_netty_Open/src/net/openmob/mobileimsdk/server/protocal/Protocal.java
/*
 * Copyright (C) 2017  即时通讯网(52im.net) & Jack Jiang.
 * The MobileIMSDK_X_netty (MobileIMSDK v3.x Netty版) Project. 
 * All rights reserved.
 * 
 * > Github地址: https://github.com/JackJiang2011/MobileIMSDK
 * > 文档地址: http://www.52im.net/forum-89-1.html
 * > 即时通讯技术社区：http://www.52im.net/
 * > 即时通讯技术交流群：320837163 (http://www.52im.net/topic-qqgroup.html)
 *  
 * "即时通讯网(52im.net) - 即时通讯开发者社区!" 推荐开源工程。
 * 
 * Protocal.java at 2017-12-9 11:24:33, code by Jack Jiang.
 * You can contact author with jack.jiang@52im.net or jb2011@163.com.
 */
package net.openmob.mobileimsdk.server.protocal;

import java.util.UUID;

import com.google.gson.Gson;

public class Protocal
{
	private boolean bridge = false;
	private int type = 0;
	private String dataContent = null;
	private String from = "-1";
	private String to = "-1";
	private String fp = null;
	private boolean QoS = false;
	private int typeu = -1;
	private transient int retryCount = 0;

	public Protocal(int type, String dataContent, String from, String to)
	{
		this(type, dataContent, from, to, -1);
	}
	
	public Protocal(int type, String dataContent, String from, String to, int typeu)
	{
		this(type, dataContent, from, to, false, null, typeu);
	}
	
	public Protocal(int type, String dataContent, String from, String to
			, boolean QoS, String fingerPrint)
	{
		this(type, dataContent, from, to, QoS, fingerPrint, -1);
	}
	
	public Protocal(int type, String dataContent, String from, String to
			, boolean QoS, String fingerPrint, int typeu)
	{
		this.type = type;
		this.dataContent = dataContent;
		this.from = from;
		this.to = to;
		this.QoS = QoS;
		this.typeu = typeu;
		
		if(QoS && fingerPrint == null)
			fp = Protocal.genFingerPrint();
		else
			fp = fingerPrint;
	}
	
	public int getType()
	{
		return this.type;
	}
	public void setType(int type)
	{
		this.type = type;
	}
	
	public String getDataContent()
	{
		return this.dataContent;
	}
	public void setDataContent(String dataContent)
	{
		this.dataContent = dataContent;
	}
	
	public String getFrom()
	{
		return this.from;
	}
	public void setFrom(String from)
	{
		this.from = from;
	}

	public String getTo()
	{
		return this.to;
	}
	public void setTo(String to)
	{
		this.to = to;
	}
	
	public String getFp()
	{
		return this.fp;
	}

	public int getRetryCount()
	{
		return this.retryCount;
	}
	public void increaseRetryCount()
	{
		this.retryCount += 1;
	}
	
	public boolean isQoS()
	{
		return QoS;
	}
	public void setQoS(boolean qoS)
	{
		this.QoS = qoS;
	}
	
	public boolean isBridge()
	{
		return bridge;
	}
	public void setBridge(boolean bridge)
	{
		this.bridge = bridge;
	}
	
	public int getTypeu()
	{
		return typeu;
	}
	public void setTypeu(int typeu)
	{
		this.typeu = typeu;
	}
	
	public String toGsonString()
	{
		return new Gson().toJson(this);
	}
	
	public byte[] toBytes()
	{
		return CharsetHelper.getBytes(toGsonString());
	}
	
	@Override
	public Object clone()
	{
		Protocal cloneP = new Protocal(this.getType()
				, this.getDataContent(), this.getFrom(), this.getTo(), this.isQoS(), this.getFp());
		cloneP.setBridge(this.bridge); // since 3.0
		cloneP.setTypeu(this.typeu);   // since 3.0
		return cloneP;
	}
	
	public static String genFingerPrint()
	{
		return UUID.randomUUID().toString();
	}
}

File: src_all/libs_src/MobileIMSDK4i_x_open/MobileIMSDK4i/Core/Protocal/Protocal.m
//  ----------------------------------------------------------------------
//  Copyright (C) 2017  即时通讯网(52im.net) & Jack Jiang.
//  The MobileIMSDK_X (MobileIMSDK v3.x) Project.
//  All rights reserved.
//
//  > Github地址: https://github.com/JackJiang2011/MobileIMSDK
//  > 文档地址: http://www.52im.net/forum-89-1.html
//  > 即时通讯技术社区：http://www.52im.net/
//  > 即时通讯技术交流群：320837163 (http://www.52im.net/topic-qqgroup.html)
//
//  "即时通讯网(52im.net) - 即时通讯开发者社区!" 推荐开源工程。
//
//  如需联系作者，请发邮件至 jack.jiang@52im.net 或 jb2011@163.com.
//  ----------------------------------------------------------------------
//
//  Protocal.m
//  MibileIMSDK4i_X (MobileIMSDK v3.0 at Summer 2017)
//
//  Created by JackJiang on 14/10/22.
//  Copyright (c) 2017年 52im.net. All rights reserved.
//

#import "Protocal.h"
#import "ToolKits.h"
#import "RMMapper.h"
#import "CharsetHelper.h"
#import "ProtocalFactory.h"


@interface Protocal ()
@property (nonatomic, assign) int retryCount;
@end


@implementation Protocal

+ (Protocal *) initWithType:(int)type content:(NSString *)dataContent from:(NSString *)from to:(NSString *)to
{
    return [Protocal initWithType:type content:dataContent from:from to:to qos:YES fp:nil tu:-1];
}
+ (Protocal *) initWithType:(int)type content:(NSString *)dataContent from:(NSString *)from to:(NSString *)to tu:(int)typeu
{
    return [Protocal initWithType:type content:dataContent from:from to:to qos:YES fp:nil tu:typeu];
}
+ (Protocal *) initWithType:(int)type content:(NSString *)dataContent from:(NSString *)from to:(NSString *)to qos:(bool)QoS fp:(NSString *)fingerPrint tu:(int)typeu
{
    return [Protocal initWithType:type content:dataContent from:from to:to qos:QoS fp:fingerPrint bg:NO tu:typeu];
}
+ (Protocal *) initWithType:(int)type content:(NSString *)dataContent from:(NSString *)from to:(NSString *)to qos:(bool)QoS fp:(NSString *)fingerPrint bg:(bool)bridge tu:(int)typeu
{
    Protocal *p = [[Protocal alloc] init];
    
    p.type = type;
    p.dataContent = dataContent;
    p.from = from;
    p.to = to;
    p.QoS = QoS;
    p.bridge = bridge;
    p.typeu = typeu;

    if(QoS && fingerPrint == nil)
        p.fp = [Protocal genFingerPrint];
    else
        p.fp = fingerPrint;
    
    return p;
}

- (id)init
{
    if(self = [super init])
    {
        self.type = 0;
        self.from = @"-1";
        self.to = @"-1";
        self.QoS = NO;
        self.bridge = NO;
        self.typeu = -1;
        self.retryCount = 0;
    }
    return self;
}

- (int) getRetryCount
{
    return self.retryCount;
}

- (void) increaseRetryCount
{
    self.retryCount += 1;
}

- (NSString *) toGsonString
{
    return [ToolKits toJSONString:[self toBytes]];
}

- (NSData *) toBytes
{
    NSMutableDictionary *dic = [self toMutableDictionary:YES];
    return [ToolKits toJSONBytesWithDictionary:dic];
}
- (NSMutableDictionary *) toMutableDictionary:(BOOL)deleteRetryCountProperty
{
    NSMutableDictionary *dic = [ToolKits toMutableDictionary:self];
    if(deleteRetryCountProperty)
        [dic removeObjectForKey:@"retryCount"];
    return dic;
}

- (Protocal *) clone
{
    NSMutableDictionary *dic = [self toMutableDictionary:YES];
    Protocal *pepFromJASON = [RMMapper objectWithClass:[Protocal class] fromDictionary:dic];
    return pepFromJASON;
}

+ (NSString *) genFingerPrint
{
    return [ToolKits generateUUID];
}

@end
